File: app/src/main/java/com/ds/rickandmorty/repository/CharacterRepository.java
package com.ds.rickandmorty.repository;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.ds.rickandmorty.model.Character;
import com.ds.rickandmorty.services.CharacterServiceProvider;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CharacterRepository {

    private static final String TAG = "char repository";
    private MutableLiveData<Character> allCharacters;
    private MutableLiveData<Character> filteredCharacters;

    public CharacterRepository() {
        allCharacters = new MutableLiveData<>();
        filteredCharacters = new MutableLiveData<>();
    }

    public void getAllCharacters(int page){
        CharacterServiceProvider.getInstance().getCharacters(page).enqueue(new Callback<Character>() {
            @Override
            public void onResponse(Call<Character> call, Response<Character> response) {
                allCharacters.postValue(response.body());
            }

            @Override
            public void onFailure(Call<Character> call, Throwable t) {
                Log.d(TAG, "onFailure: " + t.getMessage());
            }
        });
    }

    public void searchCharacters(String name, String status){
        CharacterServiceProvider.getInstance().searchCharacter(name, status).enqueue(new Callback<Character>() {
            @Override
            public void onResponse(Call<Character> call, Response<Character> response) {
                filteredCharacters.postValue(response.body());
            }

            @Override
            public void onFailure(Call<Character> call, Throwable t) {
                Log.d(TAG, "onFailure: " + t.getMessage());
            }
        });
    }

    public LiveData<Character> getAllCharLiveData(){
        return allCharacters;
    }

    public LiveData<Character> getFilteredCharLiveData(){
        return filteredCharacters;
    }
}

File: app/src/main/java/com/ds/rickandmorty/repository/CharacterRepository.kt
package com.ds.rickandmorty.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.ds.rickandmorty.model.Character
import com.ds.rickandmorty.services.CharacterServiceProvider.instanceCharService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class CharacterRepository {

    private val allCharacters: MutableLiveData<Character?> = MutableLiveData()
    private val filteredCharacters: MutableLiveData<Character?> = MutableLiveData()

    fun getAllCharacters(page: Int) {
        instanceCharService.getCharacters(page)!!.enqueue(object : Callback<Character?> {
            override fun onResponse(call: Call<Character?>, response: Response<Character?>) {
                allCharacters.postValue(response.body())
            }

            override fun onFailure(call: Call<Character?>, t: Throwable) {
                Log.d(TAG, "onFailure: " + t.message)
            }
        })
    }

    fun searchCharacters(name: String?, status: String?) {
        instanceCharService.searchCharacter(name, status)!!.enqueue(object : Callback<Character?> {
            override fun onResponse(call: Call<Character?>, response: Response<Character?>) {
                filteredCharacters.postValue(response.body())
            }

            override fun onFailure(call: Call<Character?>, t: Throwable) {
                Log.d(TAG, "onFailure: " + t.message)
            }
        })
    }

    val allCharLiveData: LiveData<Character?>
        get() = allCharacters
    val filteredCharLiveData: LiveData<Character?>
        get() = filteredCharacters

    companion object {
        private const val TAG = "char repository"
    }

}